daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/Transaction.scala

```diff
@@ -285,18 +285,16 @@ sealed abstract class HasTxNodes {
       }
     )
 
-  def byInterfaceNodes: Iterable[Node.Action] =
-    nodes.values.flatMap(node =>
-      node match {
-        case action: Node.Action =>
-          if (action.byInterface.isDefined)
-            Iterator(action)
-          else
-            Iterator()
-        case _: Node.Rollback =>
-          Iterator()
-      }
-    )
+  private[lf] def byInterfaceNodes: List[Node.Action] = {
+    val builder = List.newBuilder[Node.Action]
+    nodes.values.foreach {
+      case action: Node.Action if action.byInterface.isDefined =>
+        (builder += action)
+      case _ =>
+        ()
+    }
+    builder.result()
```

Comment:
My point was to list the nodes by traversal order, for that you can use Transaction `fold` or `foreach` method.
However, if you make it private I am completely fine you keep the iterator version.

Reply:
Since nodes is an `HashMap`, the resulting `nodes.values` will be in an unspecified order (different at each jvm restart)

Reply:
oh sorry somehow I misread this. We should definitely use `fold` or `foreach` on the transaction.

Reply:
> However, if you make it private I am completely fine you keep the original iterator version.

In fact, we should get the node by traversal order, other error message can be non deterministic.

Reply:
```suggestion
    foreach {
      case (_, action: Node.Action) if action.byInterface.isDefined =>
        (builder += action)
      case _ =>
        ()
    }
    builder.result()
```

Reply:
better to use a discard for `builder += action`

Reply:
Oh, I see! I definitely didn't understand the first time around.